commands/serve_command.go

```diff
@@ -54,6 +60,19 @@ func (c *ServeCommand) Run(args []string) int {
 		return 1
 	}
 
+	if c.cpuProfile != "" {
+		stop, err := writeCpuProfileInto(c.cpuProfile)
+		if err != nil {
+			c.Ui.Error(err.Error())
+			return 1
+		}
+		defer stop()
+	}
+
+	if c.memProfile != "" {
+		defer writeMemoryProfileInto(c.memProfile)
```

Comment:
I'm unsure how memory profiling works? It seems to me like on exit we will simply write out a memory snapshot at that moment (which is as the system is exiting) is this an accurate profile? I envision profiling to be something that is written out over time asynchronously? Perhaps a goroutine that writes out on a ticker?

I have never done this so if this is "how it's done in Go" then awesome TIL!

Reply:
I think that is mostly accurate description - i.e. yes - we would profile whatever allocated memory is left after the explicit garbage collection. It's not going to be a silver bullet for all memory-related issues, but I think it's a good start and it follows what gopls does and also what the docs suggest:

 - https://github.com/golang/tools/blob/eb789aa7ce5077c5b70da4ebd0bd36e2facf2107/go/analysis/internal/checker/checker.go#L105-L118
 - https://golang.org/pkg/runtime/pprof/

Personally I think this will likely be sufficient for the foreseeable future, because we should be far more concerned about memory leaks and "cold memory" which is still left allocated on shutdown, as opposed to short-term peaks.

I certainly not opposed to adding something more fancy as we start to care more about the runtime memory usage, but as you described this would require some more complex instrumentation and I'd like to ship something that solves the first category of problems and iterate on it later.